NHSDigital/nhsuk-react-components-extensions
# NCRS-3363: update dependencies

This is a general pull request to update dependencies to more recent versions.  The repo hasn't been updated in a few years, so some of the dependencies are no longer maintained or are incompatible with other upgrades.

**TESTING THE CHANGES**

The latest test version of these changes is published on NPM as [version 2.0.0-test.3](https://www.npmjs.com/package/nhsuk-react-components-extensions?activeTab=versions).

To check that everything works as expected when imported into another repo, I've imported that test package into two different branches of the NCRS repo:

1) Update the nhsuk-react-components-extensions-repo only, and leave all other packages the same.  See changes [here](https://github.com/NHSDigital/spine-ncrs/pull/968/files), successful pipeline [here](https://github.com/NHSDigital/spine-ncrs/actions/runs/12138554306) and [here](https://github.com/NHSDigital/spine-ncrs/actions/runs/12138554297).  

2) Update the nhsuk-react-components-extensions-repo, and also update the nhsuk-react-components from V3 to V4.  See changes [here](https://github.com/NHSDigital/spine-ncrs/pull/967/files#diff-271f950a412f5363093602cb4bf9e3dde6337a6bec3678db85a85f067b58bbbf) and passing pipelines [here](https://github.com/NHSDigital/spine-ncrs/actions/runs/12068703770) and [here](https://github.com/NHSDigital/spine-ncrs/actions/runs/12068703771).  (Note that the Typescript test step has failed - this is because the overall coverage is below the threshold, but each individual test has passed)



**LIST OF MAJOR CHANGES**

**Rollup/config changes**

Updated Rollup from V1.21 to V4.  Changed the Rollup config file from .js to .mjs format, which quieted a lot of build warnings.  Added a couple of extra plugins etc to make everything work, but otherwise no major changes.

Updated the outDir in tsconfig from `./lib` to `./dist` because I was getting compiler errors

**Test changes**

Enzyme is incompatible with React versions >16.  I have updated the tests to use Jest and react-testing-library instead.  I've tried to keep the same functionality within the tests themselves as far as possible.

**Linting**

I updated the version of eslint from 6.x to 9.x.  Version 9 onwards demands an eslint.config.mjs file, rather than an .eslintrc file like was previously being used.  

I converted the .eslintrc file into eslint.config.mjs by using @eslint/migrate-config (see discussion [here](https://eslint.org/docs/latest/use/configure/migration-guide)) and then editing the basic file it produced, so there may be some oversights.  I've had to add multiple deprecated rules to the `Rules:` section to bypass linting errors - see discussion [here](https://github.com/iamturns/eslint-config-airbnb-typescript/issues/354).

**SCSS/formatting**

The repo used node-sass for SCSS, which has since been [deprecated](https://www.npmjs.com/package/node-sass).  I swapped to Sass instead.  

Unfortunately, moving away from node-sass means that `node-sass-tilde-importer` no longer works, so the import paths have changed from nice neat ones like `~nhsuk-frontend/colours` to `../../../../node_modules/nhsuk-frontend/colours`

It looks like there's been loads of changes to the SCSS files, but it's mostly trivial refactoring, e.g. changing the notation from `color: $color_nhsuk-white` to `color: **settings**.$color_nhsuk-white`.  

I found that `tint()` and `shade()` colour mixes for the components weren't showing up properly on Storybook.  I wasn't sure whether this was a problem with my Storybook settings or the components themselves.  To be safe, I swapped `tint(colour,%)` and `shade(colour,%)` for `mix(white,colour,%)`  and `mix(black,colour,%)`, which gives the exact same effect.

I've also switched from using `@import` to `@use` as far as possible, as SCSS is planning to deprecate `@import` in the future.  (From V1.8 onwards, `@import` is completely deprecated, which means that as far as I'm aware it's not possible to use private mixins from a different stylesheet.  This is why I've kept the Sass version as ~1.70 in the package.json.)

I've tried to keep the same formatting as much as possible, although there are probably some minor changes due to the package upgrades.  Please let me know if you spot anything which looks dramatically different than before. 

**Storybook**

Upgraded various Storybook versions to get it working.  To see the different components, pull this branch and run `yarn install && yarn build && yarn storybook`

_Comparing this branch to master: I found that it wasn't possible to run Storybook in the original master branch, because it wasn't compatible with my machine._

_I've made another branch [feature/NCRS-3363_copy_of_master_with_working_storybook](https://github.com/NHSDigital/nhsuk-react-components-extensions/tree/feature/NCRS-3363_copy_of_master_with_working_storybook), which is almost identical to master but has a couple of version upgrades necessary to make Storybook work locally (or at least, they worked for me)._

_Pulling that branch and running `yarn install && yarn storybook` should hopefully bring you up a working Storybook version of the original components.  If you get an error `'ERR_OSSL_EVP_UNSUPPORTED'`, run export `NODE_OPTIONS=--openssl-legacy-provider`_


**Miscellaneous updates**

Various updates to other packages in the package.json file.

I tried updating nhsuk-frontend to the most recent version, ^9.0.0, but it gave me [pipeline failures](https://github.com/NHSDigital/nhsuk-react-components-extensions/actions/runs/12066419384/job/33647215324) since V9 is pinned to Node version >20.0.  Instead, I went for the most up-to-date version which still worked with other Node versions, which was 8.1.0

**Local FormGroup component**

We need to upgrade the NCRS repo from nhsuk-react-components V3 to nhsuk-react-components V5.  When I tried, I got weird clashes which seemed to be related to the extensions repo.  After a bit of digging, it seems like the problem might be located elsewhere, but I've come up with a workaround

The NCRS repo uses the MaskedInput component from this repo, which in turn uses the FormGroup component from nhsuk-react-components.   Whenever I tried to upgrade the components repo from V3 to V4 in NCRS, it complained that the MaskedInput component was unable to find FormGroup.  My assumption was that upgrading the extensions repo to use nhsuk-react-components V4+ would solve this problem.

I upgraded everything in this repo, and.... MaskedInput still couldn't find FormGroup.  It's related to packaging changes in the extensions repo between V3 and V4, but after digging into that repo fairly extensively, I still couldn't work out how to actually bypass the problem.

My workaround was to copy-paste FormGroup into the extensions repo, along with a few other components which it needs.  It's a bit of a hacky solution, but it solves all the problems fairly neatly - MaskedInput is unblocked, and we can now import FormGroup into NCRS directly from the extensions repo.

If anyone can suggest a better workaround to try, please let me know!  I have a good hunch for what causes the original problem; what I can't find is a solution.
